modules: openthread: add `otPlatDiagRadioTransmitCarrier`

maciejbaczmanski opened the pull request:
implemented `otPlatDiagRadioTransmitCarrier` functionality which lets user start and stop transmitting continuous carrier wave using `ot diag cw start` and `ot diag cw stop` commands.

implemented `otPlatRadioGetState` functionality to enable checking current state using `ot diag radio state` command.

disabled OT power calibration support as in Zephyr it is handled by Radio Driver, so it can't be handled on OT level.

canisLupus1313 commented on modules/openthread/platform/radio.c:
```diff
@@ -631,7 +631,7 @@ otError otPlatRadioSleep(otInstance *aInstance)
 	    sState == OT_RADIO_STATE_TRANSMIT) {
 		error = OT_ERROR_NONE;
-		if (radio_api->stop(radio_dev)) {
+		if (radio_api->stop(radio_dev) == 0) {
```
Please consider keeping consistency along all stop calls and either compare them all to 0 or remove this comparison.

maciejbaczmanski replied:
removed comparison to keep consistency with other api calls

canisLupus1313 commented on modules/openthread/platform/radio.c:
```diff
@@ -674,6 +674,36 @@ otError otPlatRadioReceiveAt(otInstance *aInstance, uint8_t aChannel,
+		}
+
+		sState = OT_RADIO_STATE_TRANSMIT;
+	} else if ((!aEnable) && (sState == OT_RADIO_STATE_TRANSMIT)) {
+		otPlatRadioReceive(aInstance, channel);
```
Why are You starting reception in transmit carrier function? And not returning any error?

maciejbaczmanski replied:
added returning of result `otPlatRadioReceive()`

rlubos commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -340,6 +340,9 @@ struct ieee802154_radio_api {
 	int (*stop)(const struct device *dev);
 
+	/** Start continuous carrier wave transmission */
+	int (*continuous_carrier)(const struct device *dev);
```
How do you leave this mode?

Plus this should be a separate commit, it's not directly related to OT.

maciejbaczmanski replied:
To leave this mode redio's state should be changed back to receive (as it is done by calling `otPlatRadioReceive()` in `platformRadioTransmitCarrier()`, which then calls `radio_api->start()`). Alternatively, changing radio's state to sleep (e.g. by `ot diag radio sleep` command) will also stop the transmission.

Moved change to a separate commit.

rlubos replied:
Yeah, but from the 15.4 radio driver perspective, I think it's not clear. IMO it should be documented, that in order to leave this mode one should call `start()` to bring it back into recv state, or `stop()`.

maciejbaczmanski replied:
Right. Added more info to comment above.

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -722,6 +722,23 @@ static int nrf5_stop(const struct device *dev)
 	return 0;
 }
 
+#if NRF_802154_CARRIER_FUNCTIONS_ENABLED
```
I think this should get a Kconfig entry in `modules/hal_nordic/Kconfig`.

Plus, a change in a specific radio driver, also merits a separate commit.

rlubos replied:
Indeed we should. Too bad I wasn't CC'ed on that.

maciejbaczmanski replied:
Moved change to separate commit

rlubos commented on modules/openthread/platform/platform-zephyr.h:
```diff
@@ -70,6 +70,11 @@ void platformUartPanic(void);
 uint16_t platformRadioChannelGet(otInstance *aInstance);
 
+/**
+ * Start/ stop continuous carrier wave transmission.
```
```suggestion
 * Start/stop continuous carrier wave transmission.
```

rlubos commented on modules/openthread/platform/radio.c:
```diff
@@ -631,9 +631,8 @@ otError otPlatRadioSleep(otInstance *aInstance)
-		if (radio_api->stop(radio_dev)) {
-			sState = OT_RADIO_STATE_SLEEP;
-		}
+		radio_api->stop(radio_dev);
+		sState = OT_RADIO_STATE_SLEEP;
```
Why this change? It's not listed in the commit message.

maciejbaczmanski replied:
Forgot to add it to the message. This fixes wrong check of `radio_api->stop()` return value
in `otPlatRadioSleep()` (api returns 0 on success, so sState was never set to `OT_RADIO_STATE_SLEEP`). 
Updated commit message.

maciejbaczmanski committed:
modules: openthread: add `otPlatDiagRadioTransmitCarrier`

added `otPlatDiagRadioTransmitCarrier` functionality

disabled OT power calibration support

fixed wrong check of `radio_api->stop()` return value
in `otPlatRadioSleep()`

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

maciejbaczmanski committed:
drivers: ieee802154: nrf5: added `continuous_carrier` API call

implemented `nrf5_continuous_carrier()` which starts continuous
carrier wave transmission

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

maciejbaczmanski committed:
net: ieee802154_radio: new API for continuous carrier wave transmission

Added new API function to start continuous carrier wave transmission

Signed-off-by: Maciej Baczmanski <maciej.baczmanski@nordicsemi.no>

rlubos reviewed (commented):
LGTM, but we should wait for https://github.com/zephyrproject-rtos/zephyr/pull/52600 and switch to use Kconfig options in the code instead.

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -724,6 +724,23 @@ static int nrf5_stop(const struct device *dev)
 	return 0;
 }
 
+#if NRF_802154_CARRIER_FUNCTIONS_ENABLED
```
Please use `#if defined(CONFIG_NRF_802154_CARRIER_FUNCTIONS)`

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -1169,6 +1186,9 @@ static struct ieee802154_radio_api nrf5_radio_api = {
 	.set_txpower = nrf5_set_txpower,
 	.start = nrf5_start,
 	.stop = nrf5_stop,
+#if NRF_802154_CARRIER_FUNCTIONS_ENABLED
```
ditto

edmont commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -724,6 +724,23 @@ static int nrf5_stop(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	if (!nrf_802154_continuous_carrier()) {
```
nit: file code style suggests bracket in new line.

edmont commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -724,6 +724,23 @@ static int nrf5_stop(const struct device *dev)
+	}
+
+	LOG_DBG("Continuous carrier wave transmission started (channel: %d)",
+		nrf_802154_channel_get());
```
nit: move to line above.
